gtk4/src/builder_rust_scope.rs

```diff
@@ -0,0 +1,143 @@
+                                .upgrade()
+                                .expect("Object destroyed before closure invalidated");
+                            let obj_v = unsafe {
+                                let mut v = glib::Value::uninitialized();
+                                glib::gobject_ffi::g_value_init(
+                                    v.to_glib_none_mut().0,
+                                    object.type_().into_glib(),
+                                );
+                                glib::gobject_ffi::g_value_set_object(
+                                    v.to_glib_none_mut().0,
+                                    object.as_object_ref().to_glib_none().0,
+                                );
+                                v
+                            };
+                            args.push(obj_v);
+                            let len = args.len();
+                            args.swap(0, len - 1);
+                            callback(&args)
+                        })
+                    } else {
+                        Closure::new_local(move |args| {
+                            let mut args = args.to_owned();
+                            let object = object_weak
+                                .upgrade()
+                                .expect("Object destroyed before closure invalidated");
+                            let obj_v = unsafe {
+                                let mut v = glib::Value::uninitialized();
+                                glib::gobject_ffi::g_value_init(
+                                    v.to_glib_none_mut().0,
+                                    object.type_().into_glib(),
+                                );
+                                glib::gobject_ffi::g_value_set_object(
+                                    v.to_glib_none_mut().0,
+                                    object.as_object_ref().to_glib_none().0,
+                                );
+                                v
+                            };
+                            args.push(obj_v);
+                            callback(&args)
+                        })
+                    };
+                    object.watch_closure(&closure);
+                    closure
+                } else {
+                    if swapped {
+                        Closure::new_local(move |args| {
+                            let mut args = args.to_owned();
+                            if !args.is_empty() {
+                                let len = args.len();
+                                args.swap(0, len - 1);
+                            }
+                            callback(&args)
+                        })
+                    } else {
+                        Closure::new_local(move |args| callback(args))
+                    }
+                }
+            })
+    }
+}
+
+impl Default for BuilderRustScope {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BuilderRustScope {
```

Comment:
```suggestion
impl RustBuilderScope {
```

Reply:
I think `RustBuilderScope` is a little easier to understand but I left it as `BuilderRustScope` for consistency with `BuilderCScope`. Maybe someone will want (unsafe) bindings for that at some point, for speed reasons? Not sure.

Reply:
The BuilderCScope is not meant for language bindings and I don't think anyone is going to write such binding, not sure if it's even possible. No strong opinions here but the RustBuilderScope is an internal thing that the user will not have to deal with it directly

Reply:
I want to leave it as `BuilderRustScope` just in case somebody wants to do something interesting by using a `Builder` and calling `set_scope` directly. I've seen some comments mentioning that `Builder` is still used occasionally.

Reply:
I added a test that shows how to use this manually with `gtk::Builder`.